### DISCUSSION TITLE
When specifying additional responses for OpenAPI, the input model is used instead of the output model

### DISCUSSION THREAD

**Discussion by jw-00000** [fastapi/fastapi#10596] [Questions] [answered]
### First Check

- [X] I added a very descriptive title here.
- [X] I used the GitHub search to find a similar question and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/pydantic/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
import pytest
from fastapi import FastAPI
from starlette.testclient import TestClient


@pytest.fixture(name="client")
def get_client():
    app = FastAPI()

    from pydantic import BaseModel, computed_field

    class Rectangle(BaseModel):
        width: int
        length: int

        @computed_field
        @property
        def area(self) -> int:
            return self.width * self.length

    # A usual endpoint, works as expected.
    @app.get("/normal")
    def normal() -> Rectangle:
        return Rectangle(width=3, length=4)

    # An endpoint with response model specified.
    @app.get("/response_model", response_model=Rectangle)
    def response_model() -> Rectangle:
        return Rectangle(width=3, length=4)

    # An endpoint with responses specified.
    @app.get("/responses", responses={200: {"model": Rectangle}})
    def responses() -> Rectangle:
        return Rectangle(width=3, length=4)

    client = TestClient(app)
    return client


def test_get(client: TestClient):
    response = client.get("/openapi.json")
    data = response.json()

    # print(data)  # Uncomment this to see the full openapi.json

    assert data["paths"]["/normal"]["get"]["responses"]["200"]["content"]["application/json"]["schema"]["$ref"] == "#/components/schemas/Rectangle-Output"
    # succeeds

    assert data["paths"]["/response_model"]["get"]["responses"]["200"]["content"]["application/json"]["schema"]["$ref"] == "#/components/schemas/Rectangle-Output"
    # succeeds

    assert data["paths"]["/responses"]["get"]["responses"]["200"]["content"]["application/json"]["schema"]["$ref"] == "#/components/schemas/Rectangle-Output"
    # fails
```

### Description

When adding the `responses` parameter to an end point, the generated openapi.json does not contain the output model, but the input model. This means that computed fields are not included in the schema.

For example, in the code above a Rectangle has a width and length, and a computed field area. The input model therefore only contains width and length, the output model also contains area.

The first end point returns a `Rectangle` as usual, and the JSON schema correctly refers to `Rectangle-Output`. (In the second end point, I add `response_model` to confirm this also works.) For the third end point, I explicitly specify that the response will contain the model Rectangle using `responses`. Surprisingly, the JSON schema now refers to `Rectangle-Input`.

Is this expected behaviour? Or am I supposed to pass the model in another way?

You can run the example code using pytest. It is based on the code from [this related question](https://github.com/tiangolo/fastapi/discussions/10342). (This issue may also be related [this solved issue](https://github.com/tiangolo/fastapi/discussions/9856).)

### Operating System

Linux

### FastAPI Version

0.104.1

### Pydantic Version

2.4.2

### Python Version

3.11.5

**Comment by jw-00000** [comment]
I think I was able to fix this problem, at least for me.

I changed [line 517 in routing.py](https://github.com/tiangolo/fastapi/blob/040ad986d48bb9a400de804f7f25abb856d85e1a/fastapi/routing.py#L517C80-L517C80) from:

```py
response_field = create_response_field(name=response_name, type_=model)
```

to:

```py
response_field = create_response_field(name=response_name, type_=model, mode="serialization")
```

I don't know what other implications this has, but it seems to work for me. I was helped by [this suggestion](https://github.com/tiangolo/fastapi/discussions/9856#discussioncomment-6636909) in issue #9856.

**Reply by flxdot** [reply, marked as answer]
Stumbled across this too a while ago and finally had the time to prepare a small PR for that: https://github.com/tiangolo/fastapi/pull/10895

**Reply by jw-00000** [reply]
Great, thanks!

**Comment by slafs** [comment]
A temporary workaround for this issue: https://github.com/fastapi/fastapi/discussions/10697#discussioncomment-10049975
